cumulusci/tasks/salesforce/metadata_etl/base.py

```diff
@@ -0,0 +1,262 @@
+    <version>{self.api_version}</version>
+</Package>
+"""
+
+    def _generate_package_xml(self, deploy):
+        return self._namespace_injector(self._get_package_xml_content(deploy))
+
+    def _create_directories(self, tempdir):
+        if self.retrieve:
+            self.retrieve_dir = Path(tempdir, "retrieve")
+            self.retrieve_dir.mkdir()
+        if self.deploy:
+            self.deploy_dir = Path(tempdir, "deploy")
+            self.deploy_dir.mkdir()
+
+    def _retrieve(self):
+        api_retrieve = ApiRetrieveUnpackaged(
+            self, self._generate_package_xml(False), self.api_version
+        )
+        unpackaged = api_retrieve()
+        unpackaged.extractall(self.retrieve_dir)
+
+    def _transform(self):
+        pass
+
+    def _deploy(self):
+        target_profile_xml = Path(self.deploy_dir, "package.xml")
+        target_profile_xml.write_text(self._generate_package_xml(True))
+
+        api = Deploy(
+            self.project_config,
+            TaskConfig(
+                {
+                    "options": {
+                        "path": self.deploy_dir,
+                        "namespace_inject": self.options.get("namespace_inject"),
+                        "unmanaged": self.options.get("unmanaged"),
+                    }
+                }
+            ),
+            self.org_config,
+        )
+        result = api()
+        self._post_deploy(result)
+
+        return result
+
+    def _post_deploy(self, result):
+        pass
+
+    def _run_task(self):
+        with tempfile.TemporaryDirectory() as tempdir:
+            self._create_directories(tempdir)
+            if self.retrieve:
+                self._retrieve()
+            self._transform()
+            if self.deploy:
+                self._deploy()
+
+
+class BaseMetadataSynthesisTask(BaseMetadataETLTask):
+    """Base class for Metadata ETL tasks that generate new metadata
+    and deploy it into the org."""
+
+    deploy = True
+
+    def _generate_package_xml(self, deploy):
+        generator = PackageXmlGenerator(str(self.deploy_dir), self.api_version)
+        return generator()
+
+    def _transform(self):
+        self._synthesize()
+
+    def _synthesize(self):
+        # Create metadata in self.deploy_dir
+        pass
+
+
+class BaseMetadataTransformTask(BaseMetadataETLTask):
+    """Base class for Metadata ETL tasks that extract metadata,
+    transform it, and deploy it back into the org."""
+
+    retrieve = True
+    deploy = True
+
+    def _get_entities(self):
+        return {}
+
+    def _get_types_package_xml(self):
+        base = """    <types>
+{members}
+        <name>{name}</name>
+    </types>
+"""
+        types = ""
+        for entity, api_names in self._get_entities().items():
+            members = "\n".join(
+                f"        <members>{api_name}</members>" for api_name in api_names
+            )
+            types += base.format(members=members, name=entity)
+
+        return types
```

Comment:
I'd sort of like to factor this out into a separate utility for generating package.xml from Python (and/or give PackageXmlGenerator a better interface for specifying types from Python rather than from components in a directory) ...but that doesn't have to happen right now.